src/tribol/interface/mfem_tribol.hpp

```diff
@@ -0,0 +1,182 @@
+ *
+ * @pre Coupling scheme cs_id must be registered using
+ * registerMfemCouplingScheme()
+ *
+ * @param [in] cs_id The ID of the coupling scheme with the MFEM mesh
+ * @param [in] v MFEM velocity ParGridFunction defined over the parent mesh
+ */
+void registerMfemVelocity( integer cs_id, const mfem::ParGridFunction& v );
+
+/**
+ * @brief Returns the response (RHS) vector to a given mfem::ParGridFunction
+ *
+ * @pre Coupling scheme cs_id must be registered using
+ * registerMfemCouplingScheme()
+ * @pre Redecomp mesh must be created and up to date by calling
+ * updateMfemParallelDecomposition()
+ * @pre Response vector must be up to date for current geometry by calling
+ * update()
+ *
+ * @param [in] cs_id The ID of the coupling scheme with the MFEM mesh
+ * @param [out] r mfem::ParGridFunction of the response (RHS) vector
+ * (pre-allocated and initialized) on the parent mesh
+ */
+void getMfemResponse( integer cs_id, mfem::ParGridFunction& r );
+
+/**
+ * @brief Get assembled contact contributions for the Jacobian matrix
+ *
+ * @pre Coupling scheme cs_id must be registered using
+ * registerMfemCouplingScheme()
+ * @pre Redecomp mesh must be created and up to date by calling
+ * updateMfemParallelDecomposition()
+ * @pre Response vector must be up to date for current geometry by calling
+ * update()
+ *
+ * This method requires registration of an mfem::ParMesh with the coupling
+ * scheme. The Jacobian contributions are split into a 2x2 block structure. The
+ * first row (or column) block is associated with the coordinate grid function
+ * degrees of freedom and the second row (or column) block is associated with
+ * the Lagrange multiplier degrees of freedom.
+ *
+ * @param csId Coupling scheme id with a registered MFEM mesh
+ * @return Jacobian contributions as an mfem::BlockOperator
+ */
+std::unique_ptr<mfem::BlockOperator> getMfemBlockJacobian( integer csId );
+
+/**
+ * @brief Returns gap vector to a given mfem::ParGridFunction
+ *
+ * @pre Coupling scheme cs_id must be registered using
+ * registerMfemCouplingScheme()
+ * @pre Redecomp mesh must be created and up to date by calling
+ * updateMfemParallelDecomposition()
+ * @pre Response vector must be up to date for current geometry by calling
+ * update()
+ *
+ * @param [in] cs_id Coupling scheme id with a registered MFEM mesh
+ * @param [out] g Nodal gap values (values do not have to be pre-allocated) on
+ * the parent-linked boundary submesh
+ */
+void getMfemGap( integer cs_id, mfem::ParGridFunction& g );
+
+/**
+ * @brief Returns reference to nodal pressure vector on the submesh surface
+ *
+ * @pre Coupling scheme cs_id must be registered using
+ * registerMfemCouplingScheme()
+ *
+ * @param cs_id Coupling scheme id with a registered MFEM mesh 
+ * @return mfem::ParGridFunction& Nodal pressure vector defined on the
+ * parent-linked boundary submesh
+ */
+mfem::ParGridFunction& getMfemPressure( integer cs_id );
```

Comment:
The pressures live on the submesh level, so that's the highest level mesh they can be retrieved on.  The API needs to expose the pressures since the host code is on the hook for solving for them and updating them.  The Jacobian matrix is also dependent on the pressures living on the submesh, since that's the FE space of the rows.

Reply:
That all makes sense. Let me go back to your patch test example to see how this might be used by a host-code in practice.